File: gcj2010R1AB.hs
#!/usr/bin/env runghc
import Data.List
import Data.List.Split
import qualified Data.Vector.Unboxed as V
main = mapM_ putStrLn . zipWith caseno [1..] . solve . lines =<< getContents
caseno i r = "Case #"++show i++": "++r
solve (c:cs) = map (toString . answer) . take (read c) . analyze $ cs ::[String]
analyze = chunksOf 2 . map (map read . words)
toString = show
answer [[d,i,m,n],a] = V.minimum $ foldl' f (V.replicate 256 0) a where
  f dp a = ins $ V.generate 256 (\x->min (del x) (mdf x)) where
    ins dp = if m==0 then dp else V.generate 256 (\x->V.minimum (V.zipWith (\v c->c+((abs (x-v))+m-1)`div`m*i) (V.enumFromN 0 256) dp))
    del x = dp V.! x + d
    mdf x = (+d) $ V.minimum $ V.slice s (e-s+1) dp where
      d = abs (a-x)
      s = max 0 (x-m)
      e = min (x+m) 255

File: gcj2010R1AB.cpp
#include <iostream>
#include <vector>
#include <algorithm>

static int solve() {
    const int VM = 256;
    int D,I,M,N;
    std::cin >> D >> I >> M >> N;
    std::vector<int> a(N+1);
    for (int i=1; i<=N; ++i) {
        std::cin >> a[i];
    }
    std::vector<std::vector<int> > dp(N+1,std::vector<int>(VM));
    for (int i=1; i<=N; ++i) {
        for (int j=0; j<VM; ++j) {
            dp[i][j] = dp[i-1][j]+D;
        }
        for (int j=0; j<VM; ++j) {
            const int s = std::max(0,j-M);
            const int e = std::min(j+M+1,VM);
            for (int k=s; k<e; ++k) {
                dp[i][j] = std::min(dp[i][j],dp[i-1][k]+std::abs(a[i]-j));
            }
        }
        if (0!=M) {
            for (int j=0; j<VM; ++j) {
                for (int k=0; k<VM; ++k) {
                    dp[i][j] = std::min(dp[i][j],dp[i][k]+(std::abs(j-k)+M-1)/M*I);
                }
            }
        }
    }
    int r = dp[N][0];
    for (int i=1; i<VM; ++i) {
        r = std::min(r,dp[N][i]);
    }
    return r;
}

int main () {
    int T;
    std::cin >> T;
    for (int i=1; i<=T; ++i) {
        std::cout << "Case #" << i << ": " << solve() << "\n";
    }
    return 0;
}
